### ISSUE TITLE
1Stop.cmd fails to identify 32-bit host when building SoftFloat-3a using VS2008

### ISSUE THREAD

**Issue by srorso** [hercules-390/SoftFloat-3a#8]
The 1Stop.cmd build script does not detect a 32-bit system when using Visual Studio 2008 because VS2008 does not set the %PLATFORM% environment variable when run on a 32-bit host system. The %PROCESSOR_ARCHITECTURE% variable does indicate 32-bit. See the comparable code in 1Stop.cmd for Hyperion.

The 1Stop.cmd also fails to recognize a requirement to build a 32-bit binary when using Visual C++ 2008 Express Edition, which only supports 32-bit targets. (Yeah, there are hacks to change this.) So, in keeping with 1Stop.cmd being a "keep it simple" tool, it will build a 32-bit binary when faced with Visual C++ 2008, even if it is a hacked Express Edition or a Standard (or better) edition that can do 64-bit targets.

See also Hyperion issue [#248](https://github.com/hercules-390/hyperion/issues/248).

**Comment by Fish-Git** [comment]
Using `%PROCESSOR_ARCHITECTURE%` _**is**_ the correct way to detect the host architecture, Steve. I've never even heard of the `%PLATFORM%` environment variable before. Is that something the newer versions of Visual Studio set?

Maybe if you could explain better what the problem is, I might be able to help you. Are you trying to build a 64-bit target on a 32-bit host? Visual Studio _**does**_ support that! (It's called cross-compiling, and VStudio has _always_ had support for that.)

Perhaps my simple [`vstools.cmd`](https://github.com/Fish-Git/telnet/blob/master/vstools.cmd) batch file may be helpful to you? (It's identical in all of my "External Package" repositories -- crypto, decNumber, SoftFloat, telnet -- so it doesn't matter which one you look at. They're all the same).

I can't remember right now for certain, but I seem to recall tripping over this same problem with CMake before myself when I was first creating my "External Package" repositories. As I recall the solution was to simply set the proper Visual Studio build environment _beforehand_ (i.e. run the `vstools.cmd` batch file), **_before_** running CMake. When you do that, then CMake does the right thing.

I hope that helps!

**Comment by Fish-Git** [comment]
> ... it will build a 32-bit binary when faced with Visual C++ 2008, even if it is a hacked Express Edition or a Standard (or better) edition that can do 64-bit targets.

That is just plain _**WRONG!**_ &nbsp; `:-O`

I'm sure if you set the proper Visual Studio build environment _before_ running CMake (perhaps via my [`vstools.cmd`](https://github.com/Fish-Git/telnet/blob/master/vstools.cmd) batch file), then CMake will very likely do the right thing!

If the user (developer) has only Visual Studio Express Edition installed (without 64-bit target support), then that's _their_ problem, _not yours!_ IMHO you should detect such a situation and throw an error instead of doing the wrong thing!!

Just my 2 cents! &nbsp; `<shrug>`

**Comment by srorso** [comment]
Hi Fish,

This issue is limited to 1Stop.cmd; CMake figures out what it can build based on specified generator and architecture options or their defaults.  CMake also complains if one requests an architecture not supported by the build tool in use.  

%PLATFORM% is set by Visual Studio 2015 and newer (and perhaps older too) to indicate the bitness of a Visual Studio command prompt build target, x86 for a 32-bit result, x64 for a 64-bit result.  I only have VS2008, which does not set %PLATFORM% and VS2015 and VS2017, which do.  

%PROCESSOR_ARCHITECTURE% indicates the bitness of the command prompt.  When a command prompt is opened from the VS2008 menu (Tools--Visual Studio 2008 Command Prompt), a 32-bit cmd.exe is started by VS2008, and %PROCESSOR_ARCHITECTURE% is set to x86 on a 64-bit system.  %PROCESSOR_ARCHITEW6432 indicates the hardware architecture.  

Use of 1Stop.cmd using Visual C++ 2008 Express Edition currently fails.  Building a 32-bit Hercules is a step forward from that.  Getting 1Stop.cmd to fully identify the capabilities of the installed VS2008 takes registry queries and coding of some complexity.  I am disinclined to do that when I am trying to replace 1Stop.cmd and the underlying NMake-based build with a CMake based build for Windows.  

(Sidebar: it was during validation of my "clean" Windows 7 and 10 virtual environments that I found this issue.  Maybe a hundred builds and I can commit CMake scripts with Windows build support.)

There is also my possibly incorrect assumption that 1Stop.cmd is intended for those folks new to building from source on Windows and just want a result without having to make the investment in becoming Windows developers.  I considered the possibility of adding `32 | 64` as an operand to 1Stop.cmd and deferred it, again because I saw that effort as a distraction and the result introduces complexity to the build process.  

Besides, if someone hacked their VS2008EE to enable 64-bit building, or has the Standard or Professional edition, then they are developers with the skills to use `makefile RETAIL-64 makefile.msvc <nn>` to build Hercules.  

Best Regards,
Steve Orso
